lib/facter/custom_facts/core/execution/base.rb

```diff
@@ -77,12 +79,31 @@ def builtin_command?(command)
           output.chomp =~ /builtin/ ? true : false
         end
 
-        def execute_command(command, on_fail, logger = nil)
+        def execute_command(command, on_fail, logger = nil, time_limit = nil)
+          @log ||= Log.new(self)
+          time_limit ||= 1.5
           begin
             # Set LC_ALL and LANG to force i18n to C for the duration of this exec;
             # this ensures that any code that parses the
             # output of the command can expect it to be in a consistent / predictable format / locale
-            out, stderr, _status_ = Open3.capture3({ 'LC_ALL' => 'C', 'LANG' => 'C' }, command.to_s)
+            opts = { 'LC_ALL' => 'C', 'LANG' => 'C' }
+            require 'timeout'
+            @log.debug("Executing command: #{command}")
+            out, stderr = Open3.popen3(opts, command.to_s) do |_, stdout, stderr, wait_thr|
+              pid = wait_thr.pid
+              output = +''
+              err = +''
+              begin
+                Timeout.timeout(time_limit) do
+                  output << stdout.read
+                  err << stderr.read
+                end
+              rescue Timeout::Error
+                @log.debug("Timeout encounter after #{time_limit}s, killing process with pid: #{pid}")
+                Process.kill('KILL', pid)
+              end
+              [output, err]
+            end
```

Comment:
Just a heads-up that `popen3` can end-up in a deadlock(as code is written now) if the command that gets executed generates to much output on stderr: https://docs.ruby-lang.org/en/2.0.0/Open3.html#method-i-popen3

> You should be careful to avoid deadlocks. Since pipes are fixed length buffer, #popen3(“prog”) {|i, o, e, t| o.read } deadlocks if the program generates many output on stderr. You should be read stdout and stderr simultaneously (using thread or IO.select). However if you don't need stderr output, #popen2 can be used. If merged stdout and stderr output is not a problem, you can use #popen2e. If you really needs stdout and stderr output as separate strings, you can consider #capture3.


I put a quick gist which i used to replicate the issue. It just outputs a lot of text to stderr: https://gist.github.com/gimmyxd/92ae2e849ecb9b1865f215853359b760

I guess this is a pretty uncommon case but it can happen.

Reply:
I don't think this is safe. If stdout/stderr are N bytes, and the child process writes N+1 bytes to `stderr`, but nothing to `stdout`, then facter will block waiting to read from `stdout`, hit the timeout, and never read from `stderr`. Meanwhile, the child process will be blocked trying to write the `N+1`th byte, which it can't until facter drains the `stderr` pipe. I think you're better off creating two reader threads, and have each append to their respective `output` and `err` buffers.